### PR TITLE
fix: Strict Projection Object Typing

### PR DESCRIPTION
**Pull request by pshaddel**
**Summary**
Fixes this issue: https://github.com/Automattic/mongoose/issues/13840

**Examples**

### CODE REVIEW

File: types/query.d.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ declare module 'mongoose' {
 
   type MongooseQueryOptions<DocType = unknown> = Pick<QueryOptions<DocType>, 'populate' | 'lean' | 'strict' | 'sanitizeProjection' | 'sanitizeFilter'>;
 
-  type ProjectionFields<DocType> = { [Key in keyof DocType]?: any } & Record<string, any>;
+  type ProjectionFields<DocType> = { [Key in keyof DocType]?: any } & Record<string, any>; // Test
```
Comment (AbdelrahmanHafez) [inline types/query.d.ts:22]:
```suggestion
  type ProjectionFields<DocType> = { [Key in keyof DocType]?: any } & Record<string, any>;
```

File: types/query.d.ts [hunk 2]
```diff
@@ -122,7 +122,7 @@ declare module 'mongoose' {
     new?: boolean;
 
     overwriteDiscriminatorKey?: boolean;
-    projection?: ProjectionType<DocType>;
+    projection?: ProjectionType<any>;
```
Comment (AbdelrahmanHafez) [inline types/query.d.ts:125]: Would that make the `projection` not take the fields in the document into account?
Is there a way to add that extra validation without using `any`?

Comment (pshaddel) [reply types/query.d.ts:125]: Thanks for pointing it out! I actually wanted to ask this. Looks like it is not used(or at least in the tests or ts-tests).

The projection has been used directly in queries:
Screenshot 2023-10-19 at 5 22 40 PM

For passing projection fields to query options we should skip the projection object:
```typescript
Model.findOne({}, {...skipProjection}, { projection: {here is the type that I changed } })
```

**Review by vkarpov15 — CHANGES_REQUESTED** [review]
A couple of use cases I'd like to see tests for that don't look like this PR supports

1. Allow adding fields that aren't in the schema to the projection
2. Projection operators: [$slice](https://www.mongodb.com/docs/manual/reference/operator/projection/slice/), [`$elemMatch`](https://www.mongodb.com/docs/manual/reference/operator/projection/elemMatch/)
3. Dotted paths, like `'child.name': 1`
4. Using `$meta` with text search: `score: { $meta: 'textScore' }`

Will have to review some more to see if there are any other projection cases that we've missed

**Comment by vkarpov15** [comment]
Another thing we'll need to make sure of: `$slice` and `$meta` are not "defining projections" in the sense that they can be used with either inclusive or exclusive projections. `{ name: 1, arr: { $slice: 5 } }` is valid as is `{ name: 0, arr: { $slice: 5 } }`.

**Comment by pshaddel** [comment]
> Another thing we'll need to make sure of: `$slice` and `$meta` are not "defining projections" in the sense that they can be used with either inclusive or exclusive projections. `{ name: 1, arr: { $slice: 5 } }` is valid as is `{ name: 0, arr: { $slice: 5 } }`.
@vkarpov15 
`$slice` and `$elemMatch` are implemented and whenever there is a array we can use these fields.

For using `$meta`, we do not know if there is a text index on the document or not; therefore, do you think we should allow this type(`{ score: { $meta: "textScore" } }`) in any projection?

**Comment by pshaddel** [comment]
> A couple of use cases I'd like to see tests for that don't look like this PR supports
> 
> 1. Allow adding fields that aren't in the schema to the projection
> 2. Projection operators: [$slice](https://www.mongodb.com/docs/manual/reference/operator/projection/slice/), [`$elemMatch`](https://www.mongodb.com/docs/manual/reference/operator/projection/elemMatch/)
> 3. Dotted paths, like `'child.name': 1`
> 4. Using `$meta` with text search: `score: { $meta: 'textScore' }`
> 
> Will have to review some more to see if there are any other projection cases that we've missed


1. resolved, we have this test:
```typsescript
Test.find({}, { someOtherField: 1 }); // This should be allowed since it's not a field in the schema
```
2. resolved
4. Dotted Path
It was quite complicated due to existence of arrays and the way we can project arrays using `.` but finally I could implement it. I had to limit the depth of the dot path to 4 because of circular types(also it could easily slow down Typescript server when the depth of the object is too much.
I also tried to support a combination of dot path and objects. For example:
```
interface X {
  docs: { users: { name: string } }[]
}

type X = Projection<X>
const x: X = { 'docs.users': { name: 1 } } // now supports typing of  this inner object and it suggests `name`
```


The next big challenge that I faced was that we are using some types, like `Date`, `Document`, `Populated`, and `Hydrated`,... that have so many methods. At the time of building the projection object, we are trying to identify the nested members of objects we might suggest fields and methods of these types as part of the projection.

Feel free to suggest your ideas and thoughts on this issue.

**Comment by fider** [comment]
@pshaddel great feature!
Will typed projection parameter also have influence on query return type?
```typescript
type EntityA = { a: number, b: ... }
const a = A.findOne({}, { a: 1 }) // typeof a == { a: number }
```

**Comment by pshaddel** [comment]
> @pshaddel great feature!
> Will typed projection parameter also have influence on query return type?
> ```typescript
> type EntityA = { a: number, b: ... }
> const a = A.findOne({}, { a: 1 }) // typeof a == { a: number }
> ```

Thanks! Still working on this feature. My intention is to implement Query Return Type in another task after tackling this one.

**Comment by vkarpov15** [comment]
@pshaddel looks like I merged a bit too early, literally the next issue I looked at highlighted an issue with this PR :disappointed: . https://github.com/Automattic/mongoose/issues/3230#issuecomment-1825839171 shows that Mongoose lets you deselect the discriminator key in an inclusive projection in addition to `_id`. What do you think about this - should we just get rid of checking for mixed inclusive/exclusive projection? Or find a way to pass in the discriminator key to this logic?

**Comment by pshaddel** [comment]
> @pshaddel looks like I merged a bit too early, literally the next issue I looked at highlighted an issue with this PR 😞 . [#3230 (comment)](https://github.com/Automattic/mongoose/issues/3230#issuecomment-1825839171) shows that Mongoose lets you deselect the discriminator key in an inclusive projection in addition to `_id`. What do you think about this - should we just get rid of checking for mixed inclusive/exclusive projection? Or find a way to pass in the discriminator key to this logic?

This should not be a problem since this discriminator key is not in the interface, so the type of the key would be `any`. it can be anything.

**Comment by vkarpov15** [comment]
That's good, but it is still valid to add `discriminatorKey` to your schema: `new Schema({ __t: String })` is fully valid.

**Comment by vkarpov15** [comment]
@pshaddel I reverted this PR in 8.1 for now with 5337da9a0f416b528dd79eb2be0c8a012865a5e8, because I don't want the test failures and other issues to block releasing 8.1. Please create a new PR with the suggested fixes; we will likely take another look at this PR for a future release as well.

**Comment by pshaddel** [comment]
OK

**Comment by alaycock** [comment]
@vkarpov15 do you know if this was ever reintroduced? I don't see a subsequent PR from @pshaddel

**Comment by pshaddel** [comment]
> @vkarpov15 do you know if this was ever reintroduced? I don't see a subsequent PR from @pshaddel
> 
> If not, maybe it's worth reopening #13840

No further PR from my side, I think a test was failing because of some other changes. I can take a look.

**Comment by pshaddel** [comment]
@alaycock Should I open a new one? I just pushed a commit to the old branch.

**Pull request merged**